### ISSUE TITLE
Loss of integer promotion when multiplying two smaller literal values into a larger type?

### ISSUE THREAD

**Issue by bbbradsmith** [cc65/cc65#1875]
I ran into a change of behaviour re: integer promotion, and I'm not sure if it's desired/intended. If I multiply two literals into something larger than an int, the don't seem to be promoted anymore, but they used to. Example code:
```
int test(unsigned long int v)
{
	if (v >= (256*256)) return 1;
	return 0;
}
```
Compiled with -O, giving the warning "Condition is always true":
```
;
; {
;
	jsr     pusheax
;
; if (v >= (256*256)) return 1;
;
	jsr     ldeax0sp
	jsr     return1
	beq     L0002
	ldx     #$00
	lda     #$01
	jmp     incsp4
;
; return 0;
;
L0002:	ldx     #$00
	txa
;
; }
;
	jmp     incsp4
```
Old output (2.17 - Git a6b04f6), no warnings:
```
;
; {
;
	jsr     pusheax
;
; if (v >= (256*256)) return 1;
;
	jsr     ldeax0sp
	cmp     #$00
	txa
	sbc     #$00
	lda     sreg
	sbc     #$01
	lda     sreg+1
	sbc     #$00
	lda     #$00
	tax
	bcc     L0001
	lda     #$01
	jmp     incsp4
;
; }
;
L0001:	jmp     incsp4
```
Two workarounds which will produce the same code as previously:
```
v >= (65536)
v >= ((unsigned long int)256*256)
```

This does give a warning, at least, so there's some diagnosis, but isn't the literal expression supposed to promote like this automatically under C's promotion rules? Was this capability removed intentionally, or was this accidental collateral of something else?

**Comment by mrdudz** [comment]
To me it looks like it is working as expected.... 256*256 gives 65536 - ie 0x10000 - which does not fit into a 16bit int, so it results in 0 and the warning is saying just that. (perhaps there could be another warning that tells about the overflow?)

**Comment by bbbradsmith** [comment]
To me it's not working as expected... code that was correctly making a 32 bit comparison before now isn't. In this case it gave a warning, but here's a related example that gives no warning at all:
```
unsigned long int test2()
{
	return 65536;
}

unsigned long int test3()
{
	return 256*256;
}
```
Result:
```
.proc	_test2: near
; return 65536;
	ldx     #$00
	stx     sreg+1
	lda     #$01
	sta     sreg
	txa
; }
	rts
.endproc

.proc	_test3: near
; return 256*256;
	ldx     #$00
	stx     sreg
	stx     sreg+1
	txa
; }
	rts
.endproc
```
Previous versions of the compiler would produce the same assembly for both of these.

Before, the automatic promotion was keeping the code generation safe, but now it silently fails. If the automatic promotion can't be done for some reason, the presence of overflow should at least be a warning (though I'd expect it to be an error).

**Comment by bbbradsmith** [comment]
One more related example:
```
unsigned long int test4()
{
	return 32768;
}

unsigned long int test5()
{
	return 32768+32768;
}
```
Compiles to:
.proc	_test4: near
; return 32768;
	ldx     #$80
	lda     #$00
	sta     sreg
	sta     sreg+1
	rts
.endproc

.proc	_test5: near
; return 32768+32768;
	ldx     #$00
	stx     sreg+1
	lda     #$01
	sta     sreg
	txa
	rts
.endproc
```
These two haven't changed between old and new versions. They are generating correct code, but give a warning on each of the "32768" literals as "Warning: Integer constant is long".

That kind of behaviour makes sense to me. It gives a warning about potentially inefficient or unintentional code, but still does the promotion and generates correct/safe code.

In the other examples above, 65536 is not identified as a long constant, and 256*256 is not promoted. Neither gives a warning, but only 65536 produces correct code, and 256*256 now silently fails. (The only reason it gave a warning in my initial example was because I was using it in a comparison, and the test for always-pass/fail comparison gave a warning.)

**Comment by mrdudz** [comment]
I agree there should be a warning.

IMHO promoting should happen only to the 16bit int type though in those cases - at least thats what i'd expect to happen in code like this. I have no idea what exactly the standard says for this case though, nor do i have the time now to read up on it :/ @acqn @spiro-trikaliotis any idea?

**Comment by bbbradsmith** [comment]
One more test, I wanted to see what happened if I keep them all in signed int range, and use only add instead of multiply:
```
unsigned long int test6()
{
	return 16384+16384+16384+16384;
}
```
New: (Silent failure, no warnings.)
```
.proc	_test6: near
; return 16384+16384+16384+16384;
	ldx     #$00
	stx     sreg
	stx     sreg+1
	txa
	rts
.endproc
```
Old: (Correct, no warnings)
```
.proc	_test6: near
; return 16384+16384+16384+16384;
	ldx     #$00
	stx     sreg+1
	lda     #$01
	sta     sreg
	txa
	rts
.endproc
```

So, this also indicates the same regression. Code is broken silently.

My feeling that "warning: integer constant is long" is good, and the compiler should be generating these warnings when you accidentally promote to a 32-bit integer without an explicit cast, but it should still produce the correct promoted result.

If it can't promote and has to overflow, that should be an error, not a warning, because the resulting code is simply wrong. It doesn't do what it looks like it does. (Though even a warning for overflow would be better than nothing.)

**Comment by mrdudz** [comment]
quick check on godbolt suggests what cc65 does is correct: https://godbolt.org/z/nqxsYbh5a

ie produce a warning, and the "broken" code

**Comment by mrdudz** [comment]
Ooops should be this one: https://godbolt.org/z/3TWzjbzvz

**Comment by bbbradsmith** [comment]
For comparison, it seems modern GCC and MSVC will give a warning of "integer constant overflow" or similar for equivalent code but for 32 to 64-bit promotion, i.e. it won't promote above a given size, but it does consistently give a warning about it.

However, the way cc65 already supports 32-bit integers, it seems very inconsistent about whether it wants to promote with a warning, promote without a warning, or overflow without a warning.

It's very easy to wrongly expect code involving 32-bit integers will work, without having used explicit casts or type suffixes. Especially since code like this has worked in previous versions of the compiler and is failing now. If the capability to automatically promote to 32-bit is being taken away, there's at least a problem here of making it clear to the user that their old code is being broken.

**Comment by bbbradsmith** [comment]
> quick check on godbolt suggests what cc65 does is correct: https://godbolt.org/z/nqxsYbh5a
> 
> ie produce a warning, and the "broken" code

In the first example (test) it's giving _a_ warning, but the wrong one. The correct warning is overflow. The constant comparison result warning is just a lucky accident.

In some of the other examples it gives no warning at all.

So, my suggestion that it should be an error, I guess that's overstepping, but what is currently happening is definitely not correct.

**Comment by mrdudz** [comment]
> Especially since code like this has worked in previous versions of the compiler and is failing now.

That btw doesnt really matter - if the compiler is doing what the standard dictates, the code was simply broken before.

**Comment by bbbradsmith** [comment]
> > Especially since code like this has worked in previous versions of the compiler and is failing now.
> That btw doesnt really matter - if the compiler is doing what the standard dictates, the code was simply broken before.

I don't think any warnings are required by the standard, but they still matter a lot.

I'm not sure if the relevant issues are directly covered by the standard. Integer sizes might be implementation-defined, and possibly their maximum automatic promotion too. At the very least, if there's something wrong with the code in the examples I provided, it's a very error-prone situation, and could probably be helped with better warnings.

I'm not saying that we have to keep the old promotion capability, if there's something that was gained by discarding it as we have, though to me it seems like a useful capability was lost. However, when there's significant changes to implementation defined behaviour, I do think it can help a lot if there's relevant warnings.

My main problem here is not that something changed about the compiler, it's that for the relevant issue the warnings are inconsistent or non-existent.

A summary of what I'm suggesting:
1. Constant arithmetic overflow doesn't seem to have any warning. Though the previous auto-promotion behaviour meant it didn't need one (at least not if it still fit in 32-bit), but if it _can't_ promote it definitely needs one.
2. "Warning: integer constant is long" seems very inconsistent. It catches 32768 but not 65536. This probably needs a review of what other cases are missing. If a literal needs a suffix or a cast to be represented properly, it should always have a warning.

**Comment by bbbradsmith** [comment]
Looking up the standards, as far as I can tell how they related to my two suggestions. Starting with #2:

Integer constants should promote to the smallest containing size starting from int, and going up to unsigned long int.

Source: https://en.cppreference.com/w/c/language/integer_constant

So, an integer constant up to 2^32-1 should by itself be valid, without needing a suffix.

A warning about long constants that don't have a suffix probably doesn't violate the standard, and it seems like a valid diagnostic for this compiler, but an implicitly long int constant shouldn't result in invalid code by itself.

The compiler looks OK in this respect, except for the warning about long integer being inconsistent.


Point #1:

As far as I can tell there aren't special rules for arithmetic with constants. So after the promotion of literals to their narrowest containing type of int or greater, I believe they follow the regular rules for arithmetic.

Rules: https://en.cppreference.com/w/c/language/operator_arithmetic

Other rules: https://en.cppreference.com/w/c/language/conversion#Integer_promotions

Unfortunately, that seems to mean that the cases I'm bringing up are _undefined_, because the offending operands would be signed int, and signed integer overflow is undefined and up to the implementation.

I believe this means that both the old and new behaviour are both standards compliant, and yes the given code is technically invalid. The implementation could automatically promote the result, it could wrap as if unsigned, or it could do anything at all. As far as I can tell, other compilers give a consistent "signed integer constant overflow" warning, and I think we should too.

Unsigned integer types on the other hand have defined wrapping to their chosen size, and arithmetic does not widen them, so for literals this is at minimum unsigned int, but could sometimes be unsigned long int (if one of the literals was big enough). This overflow could also be given a warning, but might not be appropriate.


So, I think my suggestion is mostly the same:

1. A warning on the signed int overflow is important, because it's already an unintuitive error in the 32-bit world but doubly so in 16-bit, and because it will help find new problems introduced by the migration to new behaviour. The new wrapping behaviour is OK, but it needs diagnosis.

2. The warning about long constants should either be removed (because unsuffixed long constants are valid code) or be made more consistent (because we think all long constants should be explicitly suffixed in this environment). The code behaviour of them seems correct, though.

**Comment by spiro-trikaliotis** [comment]
> A warning about long constants that don't have a suffix probably doesn't violate the standard, and it seems like a valid diagnostic for this compiler, but an implicitly long int constant shouldn't result in invalid code by itself.

You do not have an "implicitly long int constant" here. The rules of C are clear, if you multiply an int with an int, you get an int.

> As far as I can tell there aren't special rules for arithmetic with constants. So after the promotion of literals to their narrowest containing type of int or greater, I believe they follow the regular rules for arithmetic.

Yes, they do.

BTW, the syntax rules are: integer-constant -> constant -> primary-expression  -> postfix-expression -> unary-expression -> cast-expression -> multiplicative-expression.

Thus, there are not special rules for arithmetic of constants. Section 6.5.5 (N1256) then says that
>(3) The usual arithmetic conversions are performed on the operands.

> Unfortunately, that seems to mean that the cases I'm bringing up are _undefined_, because the offending operands would be signed int, and signed integer overflow is undefined and up to the implementation.

Correct.

> I believe this means that both the old and new behaviour are both standards compliant, and yes the given code is technically invalid.

Well, as the result is undefined, even the promotion to something bigger is ok, yes, this is standards compliant, as it would be compliant if the compiler started a missile to the moon.

> Unsigned integer types on the other hand have defined wrapping to their chosen size, and arithmetic does not widen them, so for literals this is at minimum unsigned int, but could sometimes be unsigned long int (if one of the literals was big enough). This overflow could also be given a warning, but might not be appropriate.

Here, we have two ints. If they were unsigned int, the result would have to be an unsigned int.

>     1. A warning on the signed int overflow is important, because it's already an unintuitive error in the 32-bit world but doubly so in 16-bit, and because it will help find new problems introduced by the migration to new behaviour. The new wrapping behaviour is OK, but it needs diagnosis.

It is arguable if it *needs* diagnostic, but it would help, yes.

>     2. The warning about long constants should either be removed (because unsuffixed long constants are valid code) or be made more consistent (because we think all long constants should be explicitly suffixed in this environment). The code behaviour of them seems correct, though.

I do not have a strong opinion on this.

**Comment by bbbradsmith** [comment]
> You do not have an "implicitly long int constant" here. The rules of C are clear, if you multiply an int with an int, you get an int.

By "implicitly" I meant that that size is determined without an explicit suffix. 100 is an int, 40000 is a long int, 2500000000 is an unsigned long int. I don't know what else to call that except its implied size, but  it seems that the spec dictates that all 3 of these should be valid code, and have an automatically determined size in this way. Is that correct?

The result of a multiply is separate, these are two different issues. Issue 2 is about the inconsistency of this warning:

```
// valid
int a = 100; // no warning
long int b = 40000; // Warning: integer constant is long
unsigned long int c = 2500000000; // no warning

// out of range
int e = 2500000000; // Warning: integer constant is unsigned long
int f = 40000; // Warning: integer constant is long
unsigned char g = 1000; // no warning
signed char h = 100; // no warning
```

So, a b and c appear to be valid code, and generate valid assembly. Why is there a warning for b and not c? What does this warning mean? In the cases where the value isn't out of range, it seems to give a warning for 32768-65535 only.

e and f get warnings but g and h don't. If it's an obtuse warning for out of range values, we seem to be missing several cases. If it's some sort of performance warning that long constants are to be avoided (?) it's being applied very inconsistently.


> Well, as the result is undefined, even the promotion to something bigger is ok, yes, this is standards compliant, as it would be compliant if the compiler started a missile to the moon.

So for issue 1, the standard doesn't dictate any sensible thing to do here, but practically I think the situation makes it very easy to accidentally build a completely invalid constant with arithmetic.

Yes it's technically invalid according to the rules, but assuming we don't live and breathe 16-bit C only, probably most of a person's intuition about what works will come from 32-bit compilers, where this particular problem hardly ever comes up because the valid ranges are much wider (and when it does, it has warnings!). Multiplying a couple of 3 digit numbers didn't set off a red flag for me, at least.

```
unsigned long i = 300 * 300; // no warning, result is truncated to 16-bit then promoted to 32-bit
```

The old behaviour of just promoting them to 32-bit seems less surprising to me than silently wrapping them. The new behaviour of wrapping them seems to have better symmetry with the unsigned case, but seems more surprising, because the code no longer does what it "looks" like it does. If we keep that behaviour, a warning would usefully compensate for the surprise.

Yes, other options would be standards compliant too, but I'm only interested in a behaviour that seems useful for programming. The compiler has to do _something_ for this case, but it doesn't have to create a silent trap under it.

**Comment by bbbradsmith** [comment]
PR #1876 adds a signed integer constant overflow warning. That gives a suitable warning on the relevant cases above.

The "integer constant is (unsigned) long" inconsistencies I don't understand yet. I'm not sure what triggers them on some cases but not others.

**Comment by rofl0r** [comment]
> long int b = 40000; // Warning: integer constant is long (?)

if you want a long integer constant, just use the L suffix, i.e. `long b = 40000L;`

**Comment by rofl0r** [comment]
> The question is why does it warn about this?

cc65's ints are 16 bit, and integers are signed, so the range is -32768 - 32767, and 40000 is clearly outside that range. integer literals without suffix are "int", not "long".

**Comment by bbbradsmith** [comment]
Per the discussion above, the standard dictates that 32768-2147483647 are long constants, and 2147483648+ is an unsigned long constant (until C99 when it became long long). These are valid constants, and they generate correct code with this compiler, in every case I've seen so far.

If the goal of the warning is to discourage the use of all long constants without an explicit suffix, then the warning has problems with consistency and should be fixed to cover all relevant cases. (i.e. I think the inconsistency is a bug.)

If there is another goal that explains the inconsistency, then it hasn't been stated yet. I feel that I'm only guessing at this warning's purpose.because what it currently does (and doesn't do) is confusing to me.

**Comment by bbbradsmith** [comment]
It was added in #1084. So the rationale is that e.g. 65535 is expected to be unsigned, so there's a warning that it's actually long and not unsigned.

I think the purpose could be clarified by changing the text of the warning. The problem is not that the constant is long, the problem is that certain ranges might be expected to be unsigned when they are not. I will make a pull request suggesting a rewording.

However, I'm finding the "unsigned long" version doesn't fire on most cases. I'm trying to figure out what's causing it to happen only sometimes.

**Comment by bbbradsmith** [comment]
I created #1877 to rephrase the two warnings in a way that I think makes it clearer what their purpose is.

Otherwise I can't seem to duplicate the problem of the "unsigned long" warning not occurring, so either I made a mistake, or the bug is really subtle, and at this point I'd rather assume I made a mistake.

**Comment by bbbradsmith** [comment]
Just to follow up on when the change of behaviour happened, I believe the change was #1322 which makes constant expression evaluation use the type sizing rules of the standard. Prior to this, I think they were always just 32-bit?

So, a good change in terms of standards compliance. I'm not complaining that it happened.

However, the consequences of turning on 16-bit constant evaluation can be a little tricky. I'm trying still trying to figure out if there are other reasonable cases that can be easily detected and might deserve a warning.

This came up for me because I have recently revisited an older project with a lot of 32-bit fixed-point stuff in it, which has offered some good test cases for the change. The signed integer overflow test definitely caught the majority of the changes, there.

**Comment by bbbradsmith** [comment]
Seems like all of the other cases I had in that program were forms of this:

```
uint32 a;
uint8 b[];
uint8 c;
a = b[c] * 256;

; old
ldy _c
ldx _b,y
lda #$00
jsr axulong
sta _a
stx _a+1
ldy sreg
sty _a+2
ldy sreg+1
sty _a+3

; new
ldy _c
ldx _b,y
lda #$00
jsr axlong
sta _a
stx _a+1
ldy sreg
sty _a+2
ldy sreg+1
sty _a+3
```

So basically it (correctly) became uint8 * sint16 = sint16 in the newer version, instead of uint8 * uint16 = uint16. I guess the old behaviour would adapt the type of a constant to the LHS result? I don't see an opportunity for a warning there, the code was just incorrect, relying on the nonstandard behaviour with no obvious way to guess that the intent was different.

Anyway, that exhausts my ideas for warnings that might help with this. The signed integer overflow detection did indicate almost all of the problems, and other compilers I checked do seem to have the same kind of warning for constants. So... I've got nothing further to suggest beyond the pull requests.
